aptos-move/framework/aptos-framework/sources/Voting.move

```diff
@@ -0,0 +1,286 @@
+        /// during execution while only relevant Table entries are.
+        proposals: Table<u64, Proposal<ProposalType>>,
+        events: VotingEvents,
+        /// Unique identifier for a proposal. This allows for 2 * 10**19 proposals.
+        next_proposal_id: u64,
+    }
+
+    struct VotingEvents has store {
+        create_proposal_events: EventHandle<CreateProposalEvent>,
+        register_forum_events: EventHandle<RegisterForumEvent>,
+        vote_events: EventHandle<VoteEvent>,
+    }
+
+    struct CreateProposalEvent has drop, store {
+        proposal_id: u64,
+        min_vote_threshold: u64,
+        expiration_secs: u64,
+    }
+
+    struct RegisterForumEvent has drop, store {
+        hosting_account: address,
+        proposal_type_info: TypeInfo,
+    }
+
+    struct VoteEvent has drop, store {
+        proposal_id: u64,
+        num_votes: u64,
+    }
+
+    public fun register<ProposalType: store>(account: &signer) {
+        let voting_forum = VotingForum<ProposalType> {
+            next_proposal_id: 0,
+            proposals: Table::new<u64, Proposal<ProposalType>>(),
+            events: VotingEvents {
+                create_proposal_events: Event::new_event_handle<CreateProposalEvent>(account),
+                register_forum_events: Event::new_event_handle<RegisterForumEvent>(account),
+                vote_events: Event::new_event_handle<VoteEvent>(account),
+            }
+        };
+
+        Event::emit_event<RegisterForumEvent>(
+            &mut voting_forum.events.register_forum_events,
+            RegisterForumEvent {
+                hosting_account: Signer::address_of(account),
+                proposal_type_info: TypeInfo::type_of<ProposalType>(),
+            },
+        );
+
+        move_to(account, voting_forum);
+    }
+
+    /// Create a proposal with the given parameters
+    ///
+    /// @param voting_forum_address The forum's address where the proposal will be stored.
+    /// @param execution_content The execution content that will be given back at resolution time. This can contain
+    /// data such as a capability resource used to scope the execution.
+    /// @param min_vote_threshold The minimum number of votes needed to consider this proposal successful.
+    /// @param expiration_secs The time in seconds at which the proposal expires and can potentially be resolved.
+    /// @return The proposal id.
+    public fun create_proposal<ProposalType: store>(
+        voting_forum_address: address,
+        execution_content: ProposalType,
+        execution_hash: Option<vector<u8>>,
+        min_vote_threshold: u64,
+        expiration_secs: u64,
+    ): u64 acquires VotingForum {
+        let voting_forum = borrow_global_mut<VotingForum<ProposalType>>(voting_forum_address);
+        let proposal_id = voting_forum.next_proposal_id;
+        voting_forum.next_proposal_id = voting_forum.next_proposal_id + 1;
```

Comment:
This should at least be gated on having some minimal stake, the creator having sufficiently long enough stake, and probably begin with the initiator locking up their stake in the vote?

Reply:
All of what you're saying is done in the separate AptosGovernance module. The idea here is (I'll put this all in a doc)

1. Voting module is only responsible for proposal creation, vote accounting, and proposal resolution.
2. Governance module that builds on top of Voting is responsible for capability control, proposal content, and for voting power calculation. There can be different approaches to prevent double voting - staking is only one of them. Another one, for example, is to use snapshots of balances. I think we should not dictate how voting power is calculated so that the Voting module can be as reusable as possible.